### PR TITLE
Add stable bond logic

### PR DESCRIPTION
**Pull request by lana-shanghai**

### CODE REVIEW

File: pallets/evercity/src/bond.rs [hunk 1]
```diff
@@ -240,31 +240,61 @@ impl<Moment, Hash> BondInnerStruct<Moment, Hash> {
...
+                && self.impact_data_baseline.iter().all(|&bl| {
+                    bl <= self.impact_data_max_deviation_cap.unwrap_or(0)
+                        && bl >= self.impact_data_max_deviation_floor.unwrap_or(0)
+                })
+        } else {
+            /// in case of the stable bond, because the below parameters are optional,
+            /// they must be None
+            self.impact_data_max_deviation_cap.is_none()
+                && self.impact_data_min_deviation_cap.is_none()
+                && self.impact_data_margin_cap..is_none()
```
Comment (ilyozzz) [inline pallets/evercity/src/bond.rs:292]: `..is_none`

File: pallets/evercity/src/bond.rs [hunk 2]
```diff
@@ -240,31 +240,61 @@ impl<Moment, Hash> BondInnerStruct<Moment, Hash> {
...
-                .bond_units_base_price
-                .saturating_mul(self.bond_units_maxcap_amount as EverUSDBalance)
-                < EverUSDBalance::MAX
-            && self.bond_duration >= MIN_BOND_DURATION
-            && self.impact_data_baseline.len() == self.bond_duration as usize
-            && self.impact_data_baseline.iter().all(|&bl| {
-                bl <= self.impact_data_max_deviation_cap
-                    && bl >= self.impact_data_max_deviation_floor
-            })
+        if !(self.is_stable()) {
```
Comment (ilyozzz) [inline pallets/evercity/src/bond.rs:254]: не стоит делать отрицание условия там, где можно написать прямую логику - читаемость снижается

```
if (self.is_stable()) {
} else {
}
```

File: pallets/evercity/src/lib.rs [hunk 3]
```diff
@@ -1877,23 +1877,28 @@ impl<T: Config> Module<T> {
...
+                    )
+                }
             } else {
-                // Report is missed, apply penalty for missed report(but not more than interest_rate_margin_cap)
-                min(
-                    bond_yields[index - 1].interest_rate
-                        + bond.inner.interest_rate_penalty_for_missed_report,
-                    bond.inner.interest_rate_margin_cap,
-                )
+                0 /// TODO - what do we return in case of the stable bond?
```
Comment (ilyozzz) [inline pallets/evercity/src/lib.rs:1901]: should return `interest_rate_base_value`

File: pallets/evercity/src/bond.rs [hunk 4]
```diff
@@ -130,60 +130,52 @@ pub struct BondInnerStruct<Moment, Hash> {
     pub impact_data_type: BondImpactType,
     /// Base value Now, all types has same interest_rate calculation logic
     /// greater the value -> lower the interest_rate and vice-versa
-    pub impact_data_baseline: Vec<u64>,
+    pub impact_data_baseline: Vec<Option<u64>>,
```
Comment (chain-police) [inline pallets/evercity/src/bond.rs:133]: Why not `Option<Vec<u64>>`?

**Review by chain-police — COMMENTED** [review]
LGTM! Just small suggestions.

There also an idea how to get rid of `unwrap_or`s. You can group the fields related to the "periodic" bond to a single struct, like so:

```rust
#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
#[derive(Encode, Decode, Clone, Default, PartialEq, RuntimeDebug)]
pub struct BondInnerStruct<Moment, Hash> {
    //... 
    interest: Option<BondInterest>
}
```

```rust
#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
#[derive(Encode, Decode, Clone, Default, PartialEq, RuntimeDebug)]
pub struct BondInterest {
    /// Base value Now, all types has same interest_rate calculation logic
    /// greater the value -> lower the interest_rate and vice-versa
    pub impact_data_baseline: Vec<u64>,

    // Coupon interest regulatory options
    /// Cap of impact_data value (absolute value). Values more then cap
    /// are considered equal to impact_data_max_deviation_cap
    /// when calculating coupon interest_rate depending on impact_data
    #[codec(compact)]
    pub impact_data_max_deviation_cap: u64,
    /// Floor of impact_data value (absolute value). Values less then floor
    /// are considered equal to impact_data_max_deviation_floor
    /// when calculating coupon interest_rate depending on impact_data
    #[codec(compact)]
    pub impact_data_max_deviation_floor: u64,
    /// Amount of seconds before end of a payment_period
    /// when Issuer should release regular impact report (confirmed by Auditor)
    #[codec(compact)]
    pub impact_data_send_period: BondPeriod,
    /// Penalty, adding to interest rate when impact report was not
    /// released during impact_data_send_period, ppm
    #[codec(compact)]
    pub interest_rate_penalty_for_missed_report: BondInterest,
    /// Base coupon interest rate, ppm. All changes of interest_rate
    /// during payment periods are based on this value, ppm
    #[codec(compact)]
    pub interest_rate_base_value: BondInterest,
    /// Upper margin of interest_rate. Interest rate cannot
    /// be more than this value, ppm
    #[codec(compact)]
    pub interest_rate_margin_cap: BondInterest,
    /// Lower margin of interest_rate. Interest rate cannot
    /// be less than this value, ppm
    #[codec(compact)]
    pub interest_rate_margin_floor: BondInterest,
    /// Interest rate during the start_periodm when interest rate is constant
    /// (from activation to first payment period), ppm
    #[codec(compact)]
    pub interest_rate_start_period_value: BondInterest,
    /// Period when Issuer should pay off coupon interests, sec
    #[codec(compact)]
    pub interest_pay_period: BondPeriod,

    /// Period from activation when effective interest rate
    /// invariably equals to interest_rate_start_period_value, sec
    #[codec(compact)]
    pub start_period: BondPeriod,
}
```

File: pallets/evercity/src/bond.rs [hunk 5]
```diff
@@ -130,60 +130,52 @@ pub struct BondInnerStruct<Moment, Hash> {
...
 
     /// <pre>
     /// This is "main" recalcualtion period of bond. Each payment_period:
     ///  - impact_data is sent to bond and confirmed by Auditor (while impact_data_send_period is active)
     ///  - coupon interest rate is recalculated for next payment_period
     ///  - required coupon interest payment is sent to bond by Issuer (while interest_pay_period is active)
     /// </pre>
-    #[codec(compact)]
-    pub payment_period: BondPeriod,
+    ///#[codec(compact)]
```
[the remark this answers is not in the document]
Comment (lana-shanghai) [reply pallets/evercity/src/bond.rs:177]: Yep, missed that - thanks!

File: pallets/evercity/src/bond.rs (hunk 4 above)
Comment (lana-shanghai) [reply pallets/evercity/src/bond.rs:133]: It was failing later in the code when getting indices from `impact_data_baseline`

Comment (chain-police) [reply pallets/evercity/src/bond.rs:133]: I guess if `bond.inner.is_stable() == false` the `impact_data_baseline` is (must be) guaranteed to be `Some`, so it should be safe to just `unwrap()` it

Comment (lana-shanghai) [reply pallets/evercity/src/bond.rs:133]: Makes sense, will try

Comment (lana-shanghai) [reply pallets/evercity/src/bond.rs:133]: No, it  fails with all the length and indices tests
`error[E0608]: cannot index into a value of type `std::option::Option<std::vec::Vec<u64>>`
    --> pallets/evercity/src/lib.rs:1891:25
     |
1891 |                         bond.inner.impact_data_baseline[index - 1].unwrap_or(0),
     |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

error[E0599]: no method named `len` found for enum `std::option::Option<std::vec::Vec<u64>>` in the current scope
   --> pallets/evercity/src/bond.rs:284:46
    |
284 |                 && self.impact_data_baseline.len() == self.bond_duration as usize
    |                                              ^^^ method not found in `std::option::Option<std::vec::Vec<u64>>``

Comment (chain-police) [reply pallets/evercity/src/bond.rs:133]: Right, here you need add a check that `self.impact_data_baseline.is_some()` and then you can safely unwrap it.

**Pull request merged**